File: src/requests/schemas/catalog.py
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field, HttpUrl, field_validator


class Category(BaseModel):
    """ข้อมูลหมวดหมู่วัสดุสำหรับแคตตาล็อก."""

    id: str = Field(..., description="รหัสหมวดหมู่ (Firestore document id)")
    name: str = Field(..., min_length=1, description="ชื่อหมวดหมู่")
    parent_id: Optional[str] = Field(
        default=None, description="รหัสหมวดหมู่แม่ หากเป็นระดับรอง"
    )
    display_order: int = Field(default=0, description="ลำดับการแสดงผลใน UI")
    thumbnail_url: Optional[HttpUrl] = Field(default=None, description="รูปภาพตัวอย่าง")
    active: bool = Field(default=True, description="สถานะเปิดใช้งาน")


class Item(BaseModel):
    """ข้อมูล SKU ที่ให้ผู้ใช้เลือกในหน้าคลังวัสดุ."""

    id: str = Field(..., description="รหัสเอกสารหรือ SKU")
    sku: str = Field(..., description="รหัส SKU ภายใน")
    name: str = Field(..., description="ชื่อวัสดุ/อุปกรณ์")
    description: Optional[str] = Field(default=None, description="รายละเอียดเพิ่มเติม")
    category_id: str = Field(..., description="รหัสหมวดหมู่ที่สินค้าอยู่")
    unit: str = Field(..., description="หน่วยนับหลัก เช่น ชิ้น, กล่อง, ม้วน")
    stock_on_hand: int = Field(..., ge=0, description="จำนวนคงเหลือพร้อมจ่าย")
    stock_reserved: int = Field(default=0, ge=0, description="จำนวนที่ถูกจองไว้")
    image_url: Optional[HttpUrl] = Field(default=None, description="ลิงก์รูปภาพจาก Cloudflare R2")
    tags: list[str] = Field(default_factory=list, description="แท็กสำหรับค้นหา")
    location_hint: Optional[str] = Field(
        default=None, description="ข้อความแนะนำตำแหน่ง (optional สำหรับ requester)"
    )
    package_width_cm: Optional[float] = Field(
        default=None, ge=0, description="ความกว้างบรรจุภัณฑ์ (เซนติเมตร)"
    )
    package_depth_cm: Optional[float] = Field(
        default=None, ge=0, description="ความลึก/ยาวบรรจุภัณฑ์ (เซนติเมตร)"
    )
    package_height_cm: Optional[float] = Field(
        default=None, ge=0, description="ความสูงบรรจุภัณฑ์ (เซนติเมตร)"
    )
    package_volume_cm3: Optional[float] = Field(
        default=None, ge=0, description="ปริมาตรต่อหน่วย (ลูกบาศก์เซนติเมตร)"
    )

    @field_validator("stock_reserved")
    @classmethod
    def validate_reserved(cls, value: int, info):  # type: ignore[override]
        if "stock_on_hand" in info.data and value > info.data["stock_on_hand"]:
            raise ValueError("stock_reserved ต้องไม่มากกว่า stock_on_hand")
        return value


class ItemCreate(BaseModel):
    """ข้อมูลที่ใช้สร้างสินค้าใหม่เข้าระบบ."""

    sku: str = Field(..., min_length=1)
    name: str = Field(..., min_length=1)
    description: Optional[str] = Field(default=None, max_length=500)
    category_id: str = Field(..., min_length=1)
    unit: str = Field(..., min_length=1)
    stock_on_hand: int = Field(default=0, ge=0)
    stock_reserved: int = Field(default=0, ge=0)
    tags: list[str] = Field(default_factory=list)
    image_url: Optional[HttpUrl] = Field(default=None)
    location_hint: Optional[str] = Field(default=None, max_length=120)
    active: bool = Field(default=True)
    package_width_cm: Optional[float] = Field(default=None, ge=0)
    package_depth_cm: Optional[float] = Field(default=None, ge=0)
    package_height_cm: Optional[float] = Field(default=None, ge=0)
    package_volume_cm3: Optional[float] = Field(default=None, ge=0)

    @field_validator("stock_reserved")
    @classmethod
    def validate_reserved(cls, value: int, info):  # type: ignore[override]
        if value < 0:
            raise ValueError("stock_reserved ต้องไม่เป็นค่าติดลบ")
        if value > info.data.get("stock_on_hand", value):
            raise ValueError("stock_reserved ต้องไม่เกิน stock_on_hand")
        return value


class ItemUpdate(BaseModel):
    """ข้อมูลที่ใช้แก้ไขสินค้า."""

    sku: Optional[str] = Field(default=None, min_length=1)
    name: Optional[str] = Field(default=None, min_length=1)
    description: Optional[str] = Field(default=None, max_length=500)
    category_id: Optional[str] = Field(default=None, min_length=1)
    unit: Optional[str] = Field(default=None, min_length=1)
    stock_on_hand: Optional[int] = Field(default=None, ge=0)
    stock_reserved: Optional[int] = Field(default=None, ge=0)
    tags: Optional[list[str]] = None
    image_url: Optional[HttpUrl] = None
    location_hint: Optional[str] = Field(default=None, max_length=120)
    active: Optional[bool] = None
    package_width_cm: Optional[float] = Field(default=None, ge=0)
    package_depth_cm: Optional[float] = Field(default=None, ge=0)
    package_height_cm: Optional[float] = Field(default=None, ge=0)
    package_volume_cm3: Optional[float] = Field(default=None, ge=0)

    @field_validator("stock_reserved")
    @classmethod
    def validate_reserved(cls, value: Optional[int], info):  # type: ignore[override]
        if value is None:
            return value
        stock_on_hand = info.data.get("stock_on_hand")
        if stock_on_hand is not None and value > stock_on_hand:
            raise ValueError("stock_reserved ต้องไม่เกิน stock_on_hand")
        return value


class CatalogSummary(BaseModel):
    """ใช้สรุปจำนวนหมวดหมู่และรายการในหน้าแรก."""

    total_categories: int = Field(..., ge=0)
    total_items: int = Field(..., ge=0)
    top_categories: list[Category] = Field(default_factory=list)

File: frontend/src/types/catalog.ts
export interface Category {
  id: string;
  name: string;
  parent_id?: string | null;
  display_order: number;
  thumbnail_url?: string | null;
  active: boolean;
}

export interface Item {
  id: string;
  sku: string;
  name: string;
  description?: string | null;
  category_id: string;
  category_type?: "material" | "equipment";
  category_name?: string | null;
  asset_code?: string | null;
  unit: string;
  stock_on_hand: number;
  stock_reserved: number;
  image_url?: string | null;
  tags: string[];
  location_hint?: string | null;
  package_width_cm?: number | null;
  package_depth_cm?: number | null;
  package_height_cm?: number | null;
  package_volume_cm3?: number | null;
}

export interface ItemCreatePayload {
  sku: string;
  name: string;
  description?: string;
  category_id?: string;
  category_name?: string;
  category_type?: "material" | "equipment";
  asset_code?: string;
  unit: string;
  stock_on_hand: number;
  stock_reserved?: number;
  tags?: string[];
  image_url?: string;
  location_hint?: string;
  active?: boolean;
  package_width_cm?: number | null;
  package_depth_cm?: number | null;
  package_height_cm?: number | null;
  package_volume_cm3?: number | null;
}

export type ItemUpdatePayload = Partial<ItemCreatePayload>;

export interface Attachment {
  key: string;
  file_name: string;
  mime_type: string;
  size_bytes: number;
  public_url?: string | null;
}

export interface RequestLinePayload {
  item_id: string;
  quantity: number;
  note?: string;
}

export interface RequestPayload {
  requester_id: string;
  requester_name: string;
  department: string;
  required_date: string;
  note?: string;
  attachments?: Attachment[];
  items: RequestLinePayload[];
}

export interface RequestResponse {
  id: string;
  requester_id: string;
  requester_name: string;
  department: string;
  required_date: string;
  status: string;
  items: Array<{
    id: string;
    item_id: string;
    quantity_requested: number;
    quantity_approved: number;
    quantity_picked: number;
    note?: string | null;
  }>;
  attachments: Attachment[];
  note?: string | null;
  created_at: string;
  updated_at: string;
}

export interface PickingQueueLine {
  request_id: string;
  request_line_id: string;
  item_id: string;
  item_name: string;
  sku: string;
  quantity_needed: number;
  quantity_allocated: number;
  required_date: string;
  note?: string | null;
  location: {
    id: string;
    zone: string;
    aisle: string;
    rack: string;
    bin_code: string;
    comment?: string | null;
  };
}

export interface PickingQueueResponse {
  total_jobs: number;
  jobs: PickingQueueLine[];
}

export type WorkOrderStatus = "planned" | "loading" | "in_transit" | "completed" | "cancelled";

export interface WorkOrderTask {
  id: string;
  request_id: string;
  request_line_id: string;
  item_id: string;
  item_name: string;
  quantity_planned: number;
  quantity_loaded: number;
  status: "pending" | "loaded" | "delivered";
}

export interface WorkOrder {
  id: string;
  route: string;
  vehicle: string;
  driver: string;
  scheduled_date: string;
  departure_time: string;
  status: WorkOrderStatus;
  tasks: WorkOrderTask[];
  created_at: string;
  updated_at: string;
}

export interface WorkOrderCreatePayload {
  route: string;
  vehicle: string;
  driver: string;
  scheduled_date: string;
  departure_time: string;
  include_request_ids?: string[];
}
